clippy_lints/src/disallowed_type.rs

```diff
@@ -0,0 +1,257 @@
+                    .collect::<Vec<_>>()
+                    .join("::");
+                span_lint(
+                    cx,
+                    DISALLOWED_TYPE,
+                    item.span,
+                    &format!("`{}` is not allowed according to config", name),
+                );
+            }
+        }
+    }
+
+    fn check_ty(&mut self, cx: &LateContext<'_>, ty: &Ty<'_>) {
+        walk_to_path(cx, ty, &self.disallowed);
+    }
+}
+
+fn walk_to_path(cx: &LateContext<'_>, ty: &Ty<'_>, disallowed: &FxHashSet<Vec<Symbol>>) {
+    match &ty.kind {
+        TyKind::Slice(t) => walk_to_path(cx, t, disallowed),
+        TyKind::Array(t, c) => {
+            walk_to_path(cx, t, disallowed);
+            // TODO: a bit hacky
+            walk_to_path(
+                cx,
+                &Ty {
+                    hir_id: ty.hir_id,
+                    kind: TyKind::Typeof(*c),
+                    span: ty.span,
+                },
+                disallowed,
+            );
+        },
+        TyKind::Ptr(mutable) => walk_to_path(cx, mutable.ty, disallowed),
+        TyKind::Rptr(_, mutable) => walk_to_path(cx, mutable.ty, disallowed),
+        TyKind::BareFn(barefn) => {
+            for input in barefn.decl.inputs {
+                walk_to_path(cx, input, disallowed);
+            }
+            if let FnRetTy::Return(t) = barefn.decl.output {
+                walk_to_path(cx, t, disallowed);
+            }
+        },
+        TyKind::Never => {},
+        TyKind::Tup(tup) => {
+            for t in *tup {
+                walk_to_path(cx, t, disallowed);
+            }
+        },
+        TyKind::Path(path) => {
+            if_chain! {
+                if let Some(did) = cx.qpath_res(path, ty.hir_id).opt_def_id();
+                let use_path = cx.get_def_path(did);
+                if let Some(name) = disallowed.iter().find(|path| **path == use_path);
+                then {
+                    let name = name.iter()
+                        .map(|s| s.to_ident_string())
+                        .collect::<Vec<_>>()
+                        .join("::");
+                    span_lint(
+                        cx,
+                        DISALLOWED_TYPE,
+                        path.span(),
+                        &format!("`{}` is not allowed according to config", name),
+                    );
+                }
+            }
+        },
+        TyKind::OpaqueDef(_, gen) => {
+            for arg in *gen {
+                if let GenericArg::Type(t) = arg {
+                    walk_to_path(cx, t, disallowed);
+                }
+            }
+        },
+        TyKind::TraitObject(polyref, _, _) => walk_trait_obj(cx, &polyref.iter().collect::<Vec<_>>(), disallowed),
+        TyKind::Typeof(anonconst) => {
+            if_chain! {
+                if let Some(tyck) = cx.maybe_typeck_results();
+                if let ty::Adt(ty::AdtDef { did, .. }, _) = tyck.node_type(anonconst.hir_id).kind();
```

Comment:
To fix ICE in CI, it would be better to use `node_type_opt`:
```suggestion
                let var_ty = tyck.node_type_opt(anonconst.hir_id);
                if let Some(var_ty) = var_ty;
                if let ty::Adt(ty::AdtDef { did, .. }, _) = var_ty.kind();
```

Reply:
Oddly enough this lint isn't affected by this, should I work on changing uses throughout the lints. It seems `node_type` is responsible for most/all of the failures?